src/Corax/Queries/UnaryMatch.Ordinal.cs

```diff
@@ -0,0 +1,248 @@
+﻿using System;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Voron;
+
+namespace Corax.Queries
+{
+    unsafe partial struct UnaryMatch<TInner, TValueType>
+    {               
+        private interface IUnaryMatchComparer
+        {
+            bool Compare(ReadOnlySpan<byte> sx, ReadOnlySpan<byte> sy);
+            bool Compare<T>(T sx, T sy) where T : unmanaged;
+        }
+
+        [SkipLocalsInit]
+        private static int AndWith(ref UnaryMatch<TInner, TValueType> match, Span<long> matches)
+        {
+            var bufferHolder = QueryContext.MatchesPool.Rent(sizeof(long) * matches.Length);
+            var buffer = MemoryMarshal.Cast<byte, long>(bufferHolder);
+
+            var count = match._fillFunc(ref match, buffer);
+            var baseMatchesPtr = (long*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(buffer));
+
+            var matchesPtr = (long*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(matches));
+            int matchesSize = matches.Length;
+
+            var result = MergeHelper.And(matchesPtr, matchesSize, matchesPtr, matchesSize, baseMatchesPtr, count);
+
+            QueryContext.MatchesPool.Return(bufferHolder);
+            return result;
+        }
+
+        [SkipLocalsInit]
+        private static int FillFuncSequence<TComparer>(ref UnaryMatch<TInner, TValueType> match, Span<long> matches)
+            where TComparer : struct, IUnaryMatchComparer
+        {
+            var currentType = ((Slice)(object)match._value).AsReadOnlySpan();
+
+            var comparer = default(TComparer);
+            var searcher = match._searcher;
+            var currentMatches = matches;
+            int totalResults = 0;
+            while (currentMatches.Length > matches.Length / 8)
```

Comment:
Why do you need `/8` space here?

Reply:
I dont want to end on a situation where we keep looping and wasting resources for very small upside. One optimization I performed that had huge impact was avoiding to go to the end in `And` operations.

Reply:
Make sense, but in this case, we need to make sure that the size is big enough to do that at least once.
Basically, ensure that we don't have sizes ` < 8`

Reply:
Good point.